src/StringOrArrayDynamicFunctionReturnTypeExtension.php

```diff
@@ -29,21 +35,63 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
     // phpcs:ignore SlevomatCodingStandard.Functions.UnusedParameter
     public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
     {
-        $args = $functionCall->getArgs();
-        if (count($args) === 0) {
+        if (count($functionCall->getArgs()) === 0) {
             return null;
         }
 
-        $dataArgType = $scope->getType($args[0]->value);
-        if ($dataArgType->isArray()->yes()) {
-            $keyType = $dataArgType->getIterableKeyType();
-            if ($keyType instanceof StringType) {
-                return new ArrayType(new StringType(), new StringType());
+        return TypeTraverser::map(
+            $scope->getType($functionCall->getArgs()[0]->value),
+            function (Type $type, callable $traverse): Type {
+                if ($type instanceof UnionType || $type instanceof IntersectionType) {
+                    // phpcs:ignore NeutronStandard.Functions.VariableFunctions.VariableFunction
+                    return $traverse($type);
+                }
+
+                return $this->getType($type);
             }
+        );
+    }
+
+    private function getType(Type $type): Type
+    {
+        if ($type->isScalar()->yes()) {
+            return new StringType();
+        }
+
+        if (!$type->isArray()->yes()) {
+            return new ConstantStringType('');
+        }
+
+        if (count($type->getConstantArrays()) > 0) {
+            return TypeCombinator::union(
```

Comment:
those wouldn't even be needed anymore I guess because this function is wrapped in a TypeTraverser already that deals with union types. the alternative would be to explicitly handle only the first constant array and array below instead of all of them. not sure what is simpler / reads better, in the end it might not matter much..

Reply:
Okay.

Reply:
@herndlm I also think that `if (count($type->getConstantArrays()) > 0) {...}` is no longer needed.

Reply:
Had a closer look. What about this?

```php
        $maybeEmptyArray = [];
        if (count($type->getConstantArrays()) > 0) {
            $constantArray = $type->getConstantArrays()[0];
            if (
                $constantArray->isOptionalKey(0) &&
                $constantArray->getArraySize()->equals(IntegerRangeType::fromInterval(0, 1))
            ) {
                $maybeEmptyArray[] = $constantArray;
            }
        }

        return TypeCombinator::union(
            ...array_merge(
                $maybeEmptyArray,
                array_map(
                    function (ArrayType $arrayType): Type {
                        return $arrayType->setOffsetValueType(
                            $arrayType->getIterableKeyType(),
                            $this->getType($arrayType->getIterableValueType())
                        );
                    },
                    $type->getArrays()
                )
            )
        );
```

Reply:
No, this fails for
```php
/** @var 'foo'|array{foo?: true, bar: int} $union */
$union = null;
assertType('array{foo?: string, bar: string}|string', esc_sql($union));
```
with
```
-'array{foo?: string, bar: string}|string'
+'array{foo?: string|true, bar: int|string}|string'
```

Reply:
got rid of the array_map and union